weather_dl/download_pipeline/manifest.py

```diff
@@ -242,22 +436,121 @@ def _update(self, download_status: DownloadStatus) -> None:
                 logger.debug(download_status)
 
 
+class BQManifest(Manifest):
+    """Writes a JSON representation of the manifest to BQ file.
+
+    This is an append-only implementation, the latest value in the manifest
+    represents the current state of a download.
+    """
+
+    _lock = threading.Lock()
+
+    def __init__(self, location: Location) -> None:
+        super().__init__(Location(location[5:]))
+        TABLE_SCHEMA = [
+            bigquery.SchemaField('selection', 'JSON', mode='REQUIRED'),
+            bigquery.SchemaField('location', 'STRING', mode='REQUIRED'),
+            bigquery.SchemaField('stage', 'STRING', mode='NULLABLE'),
+            bigquery.SchemaField('status', 'STRING', mode='REQUIRED'),
+            bigquery.SchemaField('error', 'STRING', mode='NULLABLE'),
+            bigquery.SchemaField('user', 'STRING', mode='REQUIRED'),
+            bigquery.SchemaField('size', 'FLOAT', mode='NULLABLE'),
+            bigquery.SchemaField('scheduled_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('retrieve_start_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('retrieve_end_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('fetch_start_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('fetch_end_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('download_start_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('download_end_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('upload_start_time', 'TIMESTAMP', mode='NULLABLE'),
+            bigquery.SchemaField('upload_end_time', 'TIMESTAMP', mode='NULLABLE'),
+        ]
+        table = bigquery.Table(self.location, schema=TABLE_SCHEMA)
+        with bigquery.Client() as client:
+            client.create_table(table, exists_ok=True)
+
+    def _read(self, location: str) -> DownloadStatus:
+        """Reads the JSON data from a manifest."""
+        with BQManifest._lock:
+            with bigquery.Client() as client:
+                query = f"SELECT * FROM {self.location} WHERE location = {location!r}"
+                query_job = client.query(query)
+                result = query_job.result().to_dataframe().to_dict('records')
+                row = {n: to_json_serializable_type(v) for n, v in result[0].items()}
+                return DownloadStatus.from_dict(row)
+
+    def _update(self, download_status: DownloadStatus) -> None:
+        """Writes the JSON data to a manifest."""
+        with BQManifest._lock:
```

Comment:
Yeah after thinking a bit, we actually don't need the lock in case of BQ Manifest.

Also BigQuery, does not provide traditional database transactions that follow ACID (Atomicity, Consistency, Isolation, Durability) properties.
**BUT** we are handling this by making the next transaction only when first one is complete.

Reply:
Necessary code changes done.